packages/ab-testing/src/index.ts

```diff
@@ -25,29 +28,73 @@ function getModuloValue(experiment: string, userId: number | string): number {
     return crc32.calculate(String(userId), crc32.calculate(experiment)) % 100;
 }
 
+function validateCriteria(criteria: ForceInclude, userProfile: UserProfile): boolean {
+    let itemMatchesCriteria = true;
+    for (const key in criteria) {
+        if (!criteria[key].includes(userProfile[key])) {
+            itemMatchesCriteria = false;
+        }
+    }
+
+    return itemMatchesCriteria;
+}
+
+export function validateAllocation(
+    cohort: Cohort,
+    userProfile: UserProfile,
+    userSegmentNum: number
+): boolean {
+    let withinRange = false,
+        fulfillsCriteria = true;
+    if (cohort.allocation) {
+        for (const allocation of cohort.allocation) {
+            withinRange = allocation[0] <= userSegmentNum && userSegmentNum < allocation[1];
+
+            if (withinRange) {
+                break;
+            }
+        }
+
+        if (withinRange && cohort.allocation_criteria) {
+            fulfillsCriteria = validateCriteria(cohort.allocation_criteria, userProfile);
+        }
+    }
+
+    return withinRange && fulfillsCriteria;
+}
```

Comment:
@ZigZagT , @ramon-villain , when I was refactoring this based on your comments, I realized I had an unnecessarily nested for loop when validating the allocation criteria. I moved this out and will only validate this if the withinRange check passes. Let me know what you think of this new approach!

Reply:
@ZigZagT @ramon-villain bump